### ISSUE TITLE
[FR] When using HTTPS tracking server, keep connection alive to reduce latency when submitting many metrics

### ISSUE THREAD

**Issue by iver56** [mlflow/mlflow#1867]
## Describe the proposal

When I log 50 metrics (one for each epoch in 50 epochs) using MLflow's experimental keras autolog, the time it takes to perform each request quickly adds up. It takes around 14 seconds in total my case, when communicating with an MLflow tracking server (hosted on [MFlux.ai](https://www.mflux.ai)) over HTTPS. The main reason for the slowness is that for each log_metric, a full 3-way handshake is currently performed, and this takes time. The following article explains why: https://blog.insightdatascience.com/learning-about-the-http-connection-keep-alive-header-7ebe0efa209d

What I would do is initialize an instance of requests's `Session` class when an mlflow run starts. This instance is used to keep the HTTP connection with the MLflow server alive.

Example with session, keeping the connection alive (fast):

```
s = requests.Session()
s.post("https://example.com/path/to/endpoint")
s.post("https://example.com/path/to/endpoint")
```

Example without session (slow):

```
requests.post("https://example.com/path/to/endpoint")
requests.post("https://example.com/path/to/endpoint")
```

### Motivation

The time spent waiting for metrics to be submitted to the server can be reduced by around 70%. Wouldn't it be nice to wait only 4 seconds instead of 14?

### Proposed Changes

No user-facing API changes. Only internal changes

**Comment by iver56** [comment]
In case anyone is looking for a workaround, `log_batch` is good for speeding up logging of many metrics. I wrote a quick article that shows how it can be done in mlflow 1.2: https://www.mflux.ai/tutorials/log-batch/#even-faster-example

**Comment by sueann** [comment]
@iver56 does `log_batch` satisfy your use case? it sounds like for your 50-epoch scenario you'd not want to always batch write - just in case your program crashes before the next batch write.

**Comment by iver56** [comment]
I do not think that it completely covers the use case.

I like the idea of logging keras runs with one line of code instead of ~15. Another benefit of the autolog feature is that metric timestamps are automatically set correctly, during model training. With log_batch, the default behavior is that all metrics are logged with the same timestamp.

Also, what you mentioned about crashes is sometimes relevant.

I still think that it is a good idea to reduce logging overhead by utilizing the HTTP technique suggested in this issue. 14 seconds for 50 logging calls is significant.

**Comment by chenmoneygithub** [comment]
we are working on an async logging to reduce the time overhead. for the original proposal, I am not sure if we want to hold a constant http channel until the program exits, as one server can handle multiple requests from different users.
